doc/config-generic-update-rollback/SONiC_Generic_Config_Update_and_Rollback_Design.md

```diff
@@ -0,0 +1,867 @@
+
+- Pro's
+  - Quick way to add new configuration changes
+  - It does not disrupt existing service whose configuration is not being modified. So it is non-disruptive in nature
+- Con's
+  - Can't remove existing configuration and can only be used to add/modify the existing configuration
+- Verdict
+  - Cannot be used as a rollback mechanism
+
+Since both `config load` and `config reload` are not suitable for a minimum-disruption rollback, we have to look for other approaches.
+
+In this design document, we will be exploring how to standardize the way to do partial updates, how to take checkpoints and finally how to rollback the configurations.
+
+In summary, this is the flow of an update:
+
+<img src="files/basic-target-design.png" alt="basic-target-design" width="500px"/>
+
+And the steps would be:
+```
+admin@sonic:~$ config checkpoint mycheckpoint
+admin@sonic:~$ echo "config changes to apply to ConfigDb" > some-config-changes.json-patch
+admin@sonic:~$ config apply-patch ./some-config-changes.json-patch
+admin@sonic:~$ config rollback mycheckpoint # in case of failures
+```
+
+## 1.1 Requirements
+
+### 1.1.1 Functional Requirements
+- A single, simple command to partially update SONiC configuration according to a patch of updates
+- A single, simple command to take a checkpoint of the full current SONiC config
+- A single, simple command to fully rollback current SONiC configs with to a checkpoint
+- Other commands to list checkpoints, delete checkpoints
+- The patch of updates should follow a standard notation. The [JSON Patch (RFC6902)](https://tools.ietf.org/html/rfc6902) notation should be used
+- Config rollback should be with minimum disruption to the device e.g. if reverting ACL updates DHCP should not be affected i.e. minimum-disruption rollback
+- User should be able to preview the configuration difference before going ahead and committing the configuration changes
+- In case of errors, the system should just report an error and the user should take care of it
+- Only one session globally can update device config at a time i.e. no concurrent updates to configurations
+
+### 1.1.2 Configuration and Management Requirements
+- All commands argument to generated using Python-click to provide help menus and other out-of-the box features
+- Only root user must be allowed to execute the commands
+- Non-root users can execute commands with dry-run option
+- Each command must provide the following sub options:
+  - "dry-run" Perform a dry-run of the command showing exactly what will be executed on the device, without executing it
+  - "verbose" Provide additional information on the steps executed
+
+### 1.1.3 Scalability Requirements
+N/A
+
+### 1.1.4 Warm Boot Requirements
```

Comment:
I guess, while doing a warm reboot, we put the switch in maintenance (reason: s/w upgrade) mode, and if we can detect that, and doesn't allow any config to run from the outside tool.

Reply:
My understanding is that when we make changes to ConfigDB there is 2 types of services:
- Services that need manual restart 
- Services that absorb the changes automatically

When you say "while doing a warm reboot", which one do you mean?

Reply:
Or are you saying if somehow a warm-reboot is getting executed, we should lock config changes?

Reply:
Offline discussion: The locking mechanism should be applied to config changes when doing warm-reboot, but I think this is a problem with config changes in general and not only current suggested config operation. I think it should be fixed for all operations with config-mgmt.